File: samples_c/s31_soundex.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* KAYNAK: Rosetta Code, "Soundex", C cozumu.
   https://rosettacode.org/wiki/Soundex
   (GFDL 1.2 / CC-BY-SA lisansli, atifla yeniden kullanima acik).
   add_code/init/soundex fonksiyonlari KAYNAKTAN DEGISTIRILMEDEN alinmistir;
   yalnizca main() bu deney ortaminin stdin/stdout sozlesmesine uyacak
   sekilde YENIDEN YAZILMISTIR (orijinalde sabit kodlanmis bir isim listesi
   ve karsilastirma vardi). */

/* for ASCII only */
static char code[128] = { 0 };
void add_code(const char *s, int c)
{
	while (*s) {
		code[(int)*s] = code[0x20 ^ (int)*s] = c;
		s++;
	}
}

void init(void)
{
	static const char *cls[] =
		{ "AEIOU", "", "BFPV", "CGJKQSXZ", "DT", "L", "MN", "R", 0};
	int i;
	for (i = 0; cls[i]; i++)
		add_code(cls[i], i - 1);
}

/* returns a static buffer; user must copy if want to save
   result across calls */
const char* soundex(const char *s)
{
	static char out[5];
	int c, prev, i;

	out[0] = out[4] = 0;
	if (!s || !*s) return out;

	out[0] = *s++;

	/* first letter, though not coded, can still affect next letter: Pfister */
	prev = code[(int)out[0]];
	for (i = 1; *s && i < 4; s++) {
		if ((c = code[(int)*s]) == prev) continue;

		if (c == -1) prev = 0;	/* vowel as separator */
		else if (c > 0) {
			out[i++] = c + '0';
			prev = c;
		}
	}
	while (i < 4) out[i++] = '0';
	return out;
}

int main(void)
{
	char buf[128];
	if (fgets(buf, sizeof(buf), stdin) == NULL) return 0;
	size_t len = strlen(buf);
	while (len > 0 && (buf[len-1] == '\n' || buf[len-1] == '\r')) {
		buf[--len] = '\0';
	}
	init();
	printf("%s\n", soundex(buf));
	return 0;
}

File: translations_rust_haiku/s31_soundex.rs
use std::io::{self, BufRead};

fn add_code(s: &str, c: i32, code: &mut [i32; 128]) {
    for ch in s.chars() {
        let idx = ch as usize;
        if idx < 128 {
            code[idx] = c;
            code[(0x20 ^ idx)] = c;
        }
    }
}

fn init(code: &mut [i32; 128]) {
    let cls = vec!["AEIOU", "", "BFPV", "CGJKQSXZ", "DT", "L", "MN", "R"];
    for (i, s) in cls.iter().enumerate() {
        add_code(s, (i as i32) - 1, code);
    }
}

fn soundex(s: &str, code: &[i32; 128]) -> String {
    let mut out = String::from("0000");
    if s.is_empty() {
        return out;
    }

    let mut chars = s.chars();
    if let Some(first) = chars.next() {
        out.clear();
        out.push(first);

        let mut prev = code[first as usize];
        let mut i = 1;

        for c in chars {
            if i >= 4 {
                break;
            }
            let code_val = code[c as usize];
            if code_val == prev {
                continue;
            }

            if code_val == -1 {
                prev = 0;
            } else if code_val > 0 {
                out.push((code_val as u8 + b'0') as char);
                i += 1;
                prev = code_val;
            }
        }

        while out.len() < 4 {
            out.push('0');
        }
    }

    out
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    if let Some(Ok(mut line)) = lines.next() {
        while line.ends_with('\n') || line.ends_with('\r') {
            line.pop();
        }

        let mut code = [0i32; 128];
        init(&mut code);

        println!("{}", soundex(&line, &code));
    }
}
